docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,304 @@
+requested for Cluster API on multiple occasions. This feature empowers users to reduce their
+operational resource needs, and likewise reduce their operating costs.
+
+Given that Cluster API is an abstraction point that provides access to multiple concrete cloud
+implementations, this feature might not make sense in all scenarios. To accomodate the wide
+range of deployment options in Cluster API, the scale to zero feature will be optional for
+users and infrastructure providers.
+
+### Goals
+
+- Provide capability for Cluster API MachineSets and MachineDeployments to scale from and to zero replicas.
+- Create an optional API contract in the Infrastructure Machine Template that allows infrastructure providers to specify
+  instance resource requirements that will be utilized by the cluster autoscaler.
+- Provide a mechanism for users to override the defined instance resource requirements on any given MachineSet or MachineDeployment.
+
+### Non-Goals/Future Work
+
+- Create an API contract that infrastructure providers must follow.
+- Create an API that replicates Taint and Label information from Machines to MachineSets and MachineDeployments.
+- Support for MachinePools, either with the cluster autoscaler or using infrastructure provider native implementations (eg AWS AutoScalingGroups).
+- Create an autoscaling custom resource for Cluster API.
+
+## Proposal
+
+To facilitate scaling from zero replicas, the minimal information needed by the cluster autoscaler
+is the CPU and memory resources for nodes within the target node group that will be scaled. The autoscaler
+uses this information to create a prediction about how many nodes should be created when scaling. In
+most situations this information can be directly read from the nodes that are running within a
+node group. But, during a scale from zero situation (ie when a node group has zero replicas) the
+autoscaler needs to acquire this information from the infrastructure provider.
+
+An optional status field is proposed on the Infrastructure Machine Template which will be populated
+by infrastructure providers to contain the CPU, memory, and GPU capacities for machines described by that
+template. The cluster autoscaler will then utilize this information by reading the appropriate
+infrastructure reference from the resource it is scaling (MachineSet or MachineDeployment).
+
+A user may override the field in the associated infrastructure template by applying annotations to the
+MachineSet or MachineDeployment in question. This annotation mechanism also provides users an opportunity
+to utilize scaling from zero even in situations where the infrastructure provider has not given the information
+in the Infrastructure Machine Template. In these cases the autoscaler will evaluate the annotation
+information in favor of reading the information from the status.
+
+### User Stories
+
+#### Story 1
+
+As a Cluster API user, I would like to reduce my operating costs by scaling down my workload
```

Comment:
```suggestion
As a cloud administrator, I would like to reduce my organisation's operating costs by scaling down application workloads
```

Although I realise we should add some more personas to
https://github.com/kubernetes-sigs/cluster-api/blob/58905fb5140a7844c3e24aa70d255fb8f9d0c491/docs/book/src/user/personas.md